dex/testing/firo/electrum.sh

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
+#
+# Set up Electrum-Firo (electrum) regtest client wallet for testing.
+# - Expects the firo regtest chain server harness to be running. 
+# - Expects an ElectrumX_Firo server runing on top of the chain server
+# - Connects to the ElectrumX server over SSL at localhost:EX_PORT
+# - Exposes electrum RPC at localhost:RPC_PORT - Dex connects here
+#
+# Requires:
+# - python3  - tested python3.10 and minimal testing python3.7
+# - pip3     - for boostrap loading pip
+# - git
+#
+# See Also: README_ELECTRUM_HARNESSES.md and README_HARNESS.md
+
+set -ex
+
+# https://github.com/spesmilo/electrum/issues/7833 (fixed in 4.3)
+# possible alt workaround: python3 -m pip install "protobuf>=3.12,<4"
+export PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python
+
+SCRIPT_DIR=$(pwd)
+
+# Electrum-Firo Version 4.1.5.2
+COMMIT=a3f64386efc9069cae83e23c241331de6f418b2f
+
+GENESIS=a42b98f04cc2916e8adfb5d9db8a2227c4629bc205748ed2f33180b636ee885b # regtest
+RPCPORT=8001
+EX_PORT=50002
+
+ASSET_DIR=~/dextest/electrum/firo
+ELECTRUM_DIR=${ASSET_DIR}/client
+REPO_DIR=${ELECTRUM_DIR}/electrum-repo
+WALLET_DIR=${ELECTRUM_DIR}/wallet
+NET_DIR=${WALLET_DIR}/regtest
+
+rm -rf ${NET_DIR}/blockchain_headers ${NET_DIR}/forks ${NET_DIR}/certs ${NET_DIR}/wallets/default_wallet
+mkdir -p ${NET_DIR}/regtest
+mkdir -p ${NET_DIR}/wallets
+mkdir -p ${REPO_DIR}
+
+cd ${REPO_DIR}
+
+if [ ! -d "${REPO_DIR}/.git" ]; then
+    git init
+    git remote add origin https://github.com/firoorg/electrum-firo.git
+fi
+
+git remote -v
+
+git fetch --depth 1 origin ${COMMIT}
+git reset --hard FETCH_HEAD
+
+if [ ! -d "${ELECTRUM_DIR}/venv" ]; then
+    python3 -m venv ${ELECTRUM_DIR}/venv
+fi
+source ${ELECTRUM_DIR}/venv/bin/activate
+python -m ensurepip --upgrade
+pip install -e .
+pip install requests cryptography pycryptodomex pyqt5
+
```

Comment:
Hitting a snag somewhere around here. Will try to debug but maybe you have a solution? The server seems to be running fine!
<details closed>
<summary>build failure</summary>

```
Installing collected packages: qrcode, python-bls, Pygments, protobuf, multidict, idna, frozenlist, dnspython, charset-normalizer, certifi, bitstring, attrs, async-timeout, aiorpcx, yarl, aiosignal, aiohttp, aiohttp_socks, Firo-Electrum
  DEPRECATION: python-bls is being installed using the legacy 'setup.py install' method, because it does not have a 'pyproject.toml' and the 'wheel' package is not installed. pip 23.1 will enforce this behaviour change. A possible replacement is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/8559
  Running setup.py install for python-bls ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for python-bls did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      running install
      /home/joe/dextest/electrum/firo/client/venv/lib/python3.11/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-cpython-311
      creating build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/util.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/threshold.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/tests.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/tdata.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/signature.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/pairing.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/keys.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/fields_t.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/fields.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/ec.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/bls12381.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/bls.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/aggregation_info.py -> build/lib.linux-x86_64-cpython-311/bls_py
      copying bls_py/__init__.py -> build/lib.linux-x86_64-cpython-311/bls_py
      running build_ext
      building 'bls_py.fields_t_c' extension
      creating build/temp.linux-x86_64-cpython-311
      creating build/temp.linux-x86_64-cpython-311/extmod
      creating build/temp.linux-x86_64-cpython-311/extmod/bls_py
      gcc -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -ffat-lto-objects -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection -g -ffile-prefix-map=/build/python/src=/usr/src/debug/python -flto=auto -fPIC -I/home/joe/dextest/electrum/firo/client/venv/include -I/usr/include/python3.11 -c extmod/bls_py/fields_t_c.c -o build/temp.linux-x86_64-cpython-311/extmod/bls_py/fields_t_c.o
      extmod/bls_py/fields_t_c.c: In function ‘__pyx_f_6bls_py_10fields_t_c_mpz_get_pylong_large’:
      extmod/bls_py/fields_t_c.c:2316:23: error: lvalue required as unary ‘&’ operand
       2316 |     __pyx_v_sizeptr = &Py_SIZE(((PyObject *)__pyx_v_L));
            |                       ^
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> python-bls

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
```
</details>

Reply:
Try a later / earlier version of python3 .. for me I targeted python3.10. There are some breaking changes from versions < 3.10 in the c-python headers. Not sure if this is a case like this. 
Let me know which version of python3 you are using so I can check also. I usually check against 3.7 but perhaps I did not for this script

Reply:
Any luck? .. anyway all works great my side .. Ubuntu LTS 22.04++ python3.10, pip-23.1.2-py3

And you could try playing with testnet if you like. 
![image](https://github.com/decred/dcrdex/assets/126670673/3a411baf-602f-4e78-8cdb-90dbf4a8b468)

[testnet coins here](https://testexplorer.firo.org/faucet)

Please bear in mind there is increasingly heavy testing of lelanttus spark on that testnet also right now. But you should be able to mine just using `./firo-cli  generate 1` for a while.

Reply:
`electrum.sh`

_Also tested:_
python3.7
pip-23.1.2

Builds and works fine .. but I will change the pip install line in the virtualenv

Reply:
>Hitting a snag somewhere around here. Will try to debug but maybe you have a solution? The server seems to be running fine!
>
> build failure

_Ok got deeper into the root causes:_

__python3.7__
- Works fine

__python3.10__
- Works fine

__python3.11__

- I was not even as far as you.
    -  Installed python3.11, pip, venv, etc.
    - Installed the 3.11 python C headers python3.11-dev
    - Worked through the "gmp.h" issue by installing libgmp3

- Got essentially what you got above.

So I looked on PyPI for the maintainer .. is a sometimes dash dev

[bls-python on PyPI](https://pypi.org/project/python-bls/#files)

There is a latest source. But there are _also_ a few built binary libs for various archs
- Up to and including Python 3.10
- After that nothing

__So bottom line is `python-bls` not (yet) maintained for Python > 3.10__

Please use 3.10

Reply:
<summary/>

__Good Run__

<details>

```
Initialized empty Git repository in /home/dev/dextest/electrum/firo/client/electrum-repo/.git/
+ git remote add origin https://github.com/firoorg/electrum-firo.git
+ git remote -v
origin	https://github.com/firoorg/electrum-firo.git (fetch)
origin	https://github.com/firoorg/electrum-firo.git (push)
+ git fetch --depth 1 origin a3f64386efc9069cae83e23c241331de6f418b2f
remote: Enumerating objects: 837, done.
remote: Counting objects: 100% (837/837), done.
remote: Compressing objects: 100% (772/772), done.
remote: Total 837 (delta 56), reused 486 (delta 21), pack-reused 0
Receiving objects: 100% (837/837), 7.56 MiB | 4.48 MiB/s, done.
Resolving deltas: 100% (56/56), done.
From https://github.com/firoorg/electrum-firo
 * branch            a3f64386efc9069cae83e23c241331de6f418b2f -> FETCH_HEAD
+ git reset --hard FETCH_HEAD
HEAD is now at a3f6438 Merge pull request #41 from dev-warrior777/regtest_functionality
+ '[' '!' -d /home/dev/dextest/electrum/firo/client/venv ']'
+ python3 -m venv /home/dev/dextest/electrum/firo/client/venv
+ source /home/dev/dextest/electrum/firo/client/venv/bin/activate
++ deactivate nondestructive
++ '[' -n '' ']'
++ '[' -n '' ']'
++ '[' -n /usr/bin/bash -o -n '' ']'
++ hash -r
++ '[' -n '' ']'
++ unset VIRTUAL_ENV
++ unset VIRTUAL_ENV_PROMPT
++ '[' '!' nondestructive = nondestructive ']'
++ VIRTUAL_ENV=/home/dev/dextest/electrum/firo/client/venv
++ export VIRTUAL_ENV
++ _OLD_VIRTUAL_PATH=/home/dev/.nvm/versions/node/v18.15.0/bin:/home/dev/.local/bin:/home/dev/bin:/opt/gradle/gradle-7.4.2/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/dev/bin:/usr/local/go/bin
++ PATH=/home/dev/dextest/electrum/firo/client/venv/bin:/home/dev/.nvm/versions/node/v18.15.0/bin:/home/dev/.local/bin:/home/dev/bin:/opt/gradle/gradle-7.4.2/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/dev/bin:/usr/local/go/bin
++ export PATH
++ '[' -n '' ']'
++ '[' -z '' ']'
++ _OLD_VIRTUAL_PS1=
++ PS1='(venv) '
++ export PS1
++ VIRTUAL_ENV_PROMPT='(venv) '
++ export VIRTUAL_ENV_PROMPT
++ '[' -n /usr/bin/bash -o -n '' ']'
++ hash -r
+ python --version
Python 3.10.6
+ python -m pip install --upgrade pip
Requirement already satisfied: pip in /home/dev/dextest/electrum/firo/client/venv/lib/python3.10/site-packages (22.0.2)
Collecting pip
  Using cached pip-23.1.2-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.0.2
    Uninstalling pip-22.0.2:
      Successfully uninstalled pip-22.0.2
Successfully installed pip-23.1.2
+ pip install -e .
Obtaining file:///home/dev/dextest/electrum/firo/client/electrum-repo
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Collecting qrcode (from Firo-Electrum==4.1.5.2)
  Using cached qrcode-7.4.2-py3-none-any.whl (46 kB)
Collecting protobuf>=3.12 (from Firo-Electrum==4.1.5.2)
  Using cached protobuf-4.23.4-cp37-abi3-manylinux2014_x86_64.whl (304 kB)
Collecting aiorpcx<0.19,>=0.18.7 (from Firo-Electrum==4.1.5.2)
  Using cached aiorpcX-0.18.7-py3-none-any.whl (35 kB)
Collecting aiohttp<4.0.0,>=3.3.0 (from Firo-Electrum==4.1.5.2)
  Using cached aiohttp-3.8.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.0 MB)
Collecting aiohttp-socks>=0.3 (from Firo-Electrum==4.1.5.2)
  Using cached aiohttp_socks-0.8.0-py3-none-any.whl (9.4 kB)
Collecting certifi (from Firo-Electrum==4.1.5.2)
  Using cached certifi-2023.5.7-py3-none-any.whl (156 kB)
Collecting bitstring (from Firo-Electrum==4.1.5.2)
  Using cached bitstring-4.0.2-py3-none-any.whl (46 kB)
Collecting attrs>=19.2.0 (from Firo-Electrum==4.1.5.2)
  Using cached attrs-23.1.0-py3-none-any.whl (61 kB)
Collecting x11-hash==1.4 (from Firo-Electrum==4.1.5.2)
  Using cached x11_hash-1.4.tar.gz (129 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting python-bls>=0.1.8 (from Firo-Electrum==4.1.5.2)
  Using cached python_bls-0.1.10-cp310-cp310-manylinux_2_28_x86_64.whl (844 kB)
Collecting Pygments>=2.1 (from Firo-Electrum==4.1.5.2)
  Using cached Pygments-2.15.1-py3-none-any.whl (1.1 MB)
Collecting dnspython<2.1,>=2.0 (from Firo-Electrum==4.1.5.2)
  Using cached dnspython-2.0.0-py3-none-any.whl (208 kB)
Collecting charset-normalizer<4.0,>=2.0 (from aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached charset_normalizer-3.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (201 kB)
Collecting multidict<7.0,>=4.5 (from aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached multidict-6.0.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (114 kB)
Collecting async-timeout<5.0,>=4.0.0a3 (from aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
Collecting yarl<2.0,>=1.0 (from aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached yarl-1.9.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (268 kB)
Collecting frozenlist>=1.1.1 (from aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached frozenlist-1.3.3-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (149 kB)
Collecting aiosignal>=1.1.2 (from aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached aiosignal-1.3.1-py3-none-any.whl (7.6 kB)
Collecting python-socks[asyncio]<3.0.0,>=2.0.0 (from aiohttp-socks>=0.3->Firo-Electrum==4.1.5.2)
  Using cached python_socks-2.3.0-py3-none-any.whl (53 kB)
Collecting typing-extensions (from qrcode->Firo-Electrum==4.1.5.2)
  Using cached typing_extensions-4.7.1-py3-none-any.whl (33 kB)
Collecting pypng (from qrcode->Firo-Electrum==4.1.5.2)
  Using cached pypng-0.20220715.0-py3-none-any.whl (58 kB)
Collecting idna>=2.0 (from yarl<2.0,>=1.0->aiohttp<4.0.0,>=3.3.0->Firo-Electrum==4.1.5.2)
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Building wheels for collected packages: Firo-Electrum, x11-hash
  Building editable for Firo-Electrum (pyproject.toml) ... done
  Created wheel for Firo-Electrum: filename=Firo_Electrum-4.1.5.2-0.editable-py3-none-any.whl size=10690 sha256=a3b8f01a03b1a4a7db4fd641c2aebf1be0a206deda3e7fa7fa9216633c402c06
  Stored in directory: /tmp/pip-ephem-wheel-cache-qo7jl22q/wheels/4e/46/24/2253f986cb4f6b51edfa50129f521b00afe526c649f276d3e4
  Building wheel for x11-hash (pyproject.toml) ... done
  Created wheel for x11-hash: filename=x11_hash-1.4-cp310-cp310-linux_x86_64.whl size=563333 sha256=4559c6946b3bae90b1f3e4020e32004a374ed6de2130ff4268e09d1035b84f93
  Stored in directory: /home/dev/.cache/pip/wheels/72/7e/ee/f1381c49bc04080481e5f8ce83c867d374d308f00e5788b021
Successfully built Firo-Electrum x11-hash
Installing collected packages: x11-hash, python-socks, pypng, typing-extensions, python-bls, Pygments, protobuf, multidict, idna, frozenlist, dnspython, charset-normalizer, certifi, bitstring, attrs, async-timeout, aiorpcx, yarl, qrcode, aiosignal, aiohttp, aiohttp-socks, Firo-Electrum
Successfully installed Firo-Electrum-4.1.5.2 Pygments-2.15.1 aiohttp-3.8.4 aiohttp-socks-0.8.0 aiorpcx-0.18.7 aiosignal-1.3.1 async-timeout-4.0.2 attrs-23.1.0 bitstring-4.0.2 certifi-2023.5.7 charset-normalizer-3.2.0 dnspython-2.0.0 frozenlist-1.3.3 idna-3.4 multidict-6.0.4 protobuf-4.23.4 pypng-0.20220715.0 python-bls-0.1.10 python-socks-2.3.0 qrcode-7.4.2 typing-extensions-4.7.1 x11-hash-1.4 yarl-1.9.2
+ pip install requests cryptography pycryptodomex pyqt5
Collecting requests
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Collecting cryptography
  Using cached cryptography-41.0.1-cp37-abi3-manylinux_2_28_x86_64.whl (4.3 MB)
Collecting pycryptodomex
  Using cached pycryptodomex-3.18.0-cp35-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.1 MB)
Collecting pyqt5
  Using cached PyQt5-5.15.9-cp37-abi3-manylinux_2_17_x86_64.whl (8.4 MB)
Requirement already satisfied: charset-normalizer<4,>=2 in /home/dev/dextest/electrum/firo/client/venv/lib/python3.10/site-packages (from requests) (3.2.0)
Requirement already satisfied: idna<4,>=2.5 in /home/dev/dextest/electrum/firo/client/venv/lib/python3.10/site-packages (from requests) (3.4)
Collecting urllib3<3,>=1.21.1 (from requests)
  Using cached urllib3-2.0.3-py3-none-any.whl (123 kB)
Requirement already satisfied: certifi>=2017.4.17 in /home/dev/dextest/electrum/firo/client/venv/lib/python3.10/site-packages (from requests) (2023.5.7)
Collecting cffi>=1.12 (from cryptography)
  Using cached cffi-1.15.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (441 kB)
Collecting PyQt5-sip<13,>=12.11 (from pyqt5)
  Using cached PyQt5_sip-12.12.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.whl (360 kB)
Collecting PyQt5-Qt5>=5.15.2 (from pyqt5)
  Using cached PyQt5_Qt5-5.15.2-py3-none-manylinux2014_x86_64.whl (59.9 MB)
Collecting pycparser (from cffi>=1.12->cryptography)
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: PyQt5-Qt5, urllib3, PyQt5-sip, pycryptodomex, pycparser, requests, pyqt5, cffi, cryptography
Successfully installed PyQt5-Qt5-5.15.2 PyQt5-sip-12.12.1 cffi-1.15.1 cryptography-41.0.1 pycparser-2.21 pycryptodomex-3.18.0 pyqt5-5.15.9 requests-2.31.0 urllib3-2.0.3
+ cp /home/dev/dex/dcrdex/dex/testing/firo/electrum_regtest_wallet /home/dev/dextest/electrum/firo/client/wallet/regtest/wallets/default_wallet
+ cat
+ cat
+ ./electrum-firo --regtest --dir=/home/dev/dextest/electrum/firo/client/wallet
```
</details>